scripts/map.js

```diff
@@ -0,0 +1,135 @@
+(function(module) {
+  mapObj = {};
+  // mapObj.geoCodeJSON = [];
+  // mapObj.parsedLocation = [];
+  mapObj.markers = [];
+  mapObj.dbData = [];
+  mapObj.lastOpenInfoWindow;
+//** dont forget to delete
+  // mapObj.dummyData = ['afghanistan', 'iraq', 'india'];
+// get visitor location data and query google api
+  mapObj.googleReq = function(callback) {
+    this.mapRender();
+    $.get('/location', function(data) {
+      // instead of separate arrays, I am mapping the new data and including the
+      // the original properties--that way it will be easier to post the data
+      // back to the database
+      mapObj.dbData = data.map(function(d) {
+        // google gets confused about "Washington"--defaults to Washington DC
+        // the geoAddress is used to query the geocoder
+        if (d.country_answer) {
+          d.geoAddress = 'country+=+' + d.country_answer.replace(/\s+/g, '+');
+        } else if (d.state_answer) {
+          d.geoAddress = 'state+=+' + d.state_answer.replace(/\s+/g, '+');
+        } else {
+          console.error('this is bad data');
+        }
+        return d;
+      });
+    }).then(function() {
+      var i = 0;
+      mapObj.dbData.forEach(function(location) {
+        /**
+        TO DO: check if geoData already exists on each location
+          if geoData exists, then addMarker
+          else, query geocoder
+        **/
+        if(location.lat && location.lng) {
+          // console.log(geoData);
+          mapObj.addMarker(location, 'db');
+        } else {
+          // because Google gets mad: the i multiplies the delay by every iteration
+          i += 1;
+          setTimeout(function() {
+            mapObj.geocoder.geocode({'address': location.geoAddress}, function(results, status) {
+              if (status === 'OK') {
```

Comment:
There's quite a bit going on here with nested functions and conditionals. Might warrant some helper functions.